File: src/main/services/xml-utils.ts
/**
 * Shared XML utilities — escaping, entity decoding, and language mapping.
 *
 * Extracted from xmp-inject.ts so both the XMP path and the ComicInfo path
 * use the same escaping/decoding logic without duplication.
 */

// ─── Entities (built by string concatenation to prevent source corruption) ───

const AMP = '&' + 'amp;'
const LT = '&' + 'lt;'
const GT = '&' + 'gt;'
const QUOT = '&' + 'quot;'
const APOS = '&#' + '39;'

// ─── Escaping ─────────────────────────────────────────────────────────────────

/** Characters that are illegal in XML 1.0 even when escaped. */
// eslint-disable-next-line no-control-regex
const ILLEGAL_XML_CHARS = /[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]/g

/**
 * Escape text for inclusion in XML element text or attribute values.
 *
 * Each entity replacement is built via string concatenation so the source
 * file cannot accidentally contain entity-like text that gets mangled.
 */
export function escapeXml(s: string): string {
  return s
    .replace(ILLEGAL_XML_CHARS, '')
    .replace(/&/g, AMP)
    .replace(/</g, LT)
    .replace(/>/g, GT)
    .replace(/"/g, QUOT)
    .replace(/'/g, APOS)
}

/**
 * Decode XML character entities back to their literal characters.
 *
 * This is the inverse of escapeXml(). Must decode `&` last to avoid
 * double-decoding `&lt;` -> `<` instead of `<` -> `<`.
 *
 * Uses the same concatenation trick so the entity patterns are real,
 * not raw characters invisible in the source.
 */
export function decodeXmlEntities(s: string): string {
  return s
    .replace(/&#x([0-9a-fA-F]+);/g, (_, hex: string) => String.fromCodePoint(parseInt(hex, 16)))
    .replace(/&#(\d+);/g, (_, dec: string) => String.fromCodePoint(parseInt(dec, 10)))
    .replace(new RegExp(escapeRegex(LT), 'g'), '<')
    .replace(new RegExp(escapeRegex(GT), 'g'), '>')
    .replace(new RegExp(escapeRegex(QUOT), 'g'), '"')
    .replace(new RegExp(escapeRegex(APOS), 'g'), "'")
    .replace(new RegExp(escapeRegex(AMP), 'g'), '&') // must come last
}

/** Escape special regex characters in a string so it can be used as a literal pattern. */
function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

// ─── Language ─────────────────────────────────────────────────────────────────

/**
 * Map language values seen in this library to ISO 639-1 codes.
 *
 * Three sources feed this field, and they do not agree:
 * - nhentai tag names, which are English words (`english`, `japanese`)
 * - `library_item.custom_language`, which the scanner reads out of existing PDF
 *   XMP and which holds **ISO 639-2 three-letter codes** (`eng`, `jpn`, `zho`)
 * - hand-edited values, which are already two-letter codes
 *
 * The three-letter forms are the majority case, not an edge case: measured on
 * this library, 2,390 rows read `eng`, 398 `jpn` and 221 `zho` — together over
 * 60% of the collection. Omitting them meant most converted files carried no
 * language at all.
 */
const LANGUAGE_TO_ISO: Record<string, string> = {
  // nhentai tag names
  english: 'en',
  japanese: 'ja',
  chinese: 'zh',
  korean: 'ko',
  french: 'fr',
  spanish: 'es',
  german: 'de',
  italian: 'it',
  portuguese: 'pt',
  russian: 'ru',
  other: 'ot',
  // ISO 639-2 three-letter codes, as written into PDF XMP. Both the
  // terminological (T) and bibliographic (B) forms appear in the wild, so map
  // both where they differ: zho/chi, deu/ger, fra/fre.
  eng: 'en',
  jpn: 'ja',
  zho: 'zh',
  chi: 'zh',
  kor: 'ko',
  fra: 'fr',
  fre: 'fr',
  spa: 'es',
  deu: 'de',
  ger: 'de',
  ita: 'it',
  por: 'pt',
  rus: 'ru'
}

/**
 * Convert a free-text language value to an ISO 639-1 code.
 *
 * Returns `null` if the input is empty, unrecognisable, or not a valid language
 * (e.g. the free-text value `translated`).
 */
export function toIsoLanguage(lang: string | null | undefined): string | null {
  if (!lang) return null
  const lower = lang.toLowerCase().trim()
  if (!lower) return null
  if (/^[a-z]{2}$/.test(lower)) return lower
  return LANGUAGE_TO_ISO[lower] || null
}

/**
 * The only language values this app stores. Order is priority order.
 *
 * A gallery is one language for our purposes, so a translated Japanese work
 * tagged `japanese, translated, english` resolves to English: that is the
 * language of the text a reader will actually see.
 *
 * Measured on this library, the API only ever returns these four
 * `language`-type tag values — `translated` (27), `english` (20),
 * `japanese` (15), `chinese` (9) — in these combinations:
 *
 *   ('translated', 'english')   18    ('japanese',)   15
 *   ('translated', 'chinese')    9    ('english',)     2
 *
 * Note that `translated` comes *first* whenever it is present, which is why
 * taking the first language tag mislabelled the majority of downloads.
 */
export const CANONICAL_LANGUAGES = ['English', 'Japanese', 'Chinese'] as const

export type CanonicalLanguage = (typeof CANONICAL_LANGUAGES)[number]

/**
 * Every spelling seen for each canonical language: nhentai tag names, ISO 639-1
 * codes, and the ISO 639-2 codes the scanner reads out of existing PDF XMP.
 */
const LANGUAGE_ALIASES: Record<CanonicalLanguage, string[]> = {
  English: ['english', 'en', 'eng'],
  Japanese: ['japanese', 'ja', 'jpn', 'jp'],
  Chinese: ['chinese', 'zh', 'zho', 'chi', 'cn']
}

/**
 * Resolve a gallery's language tags to exactly one canonical name.
 *
 * Values that are not languages — `translated`, `rewrite`, `speechless` — match
 * nothing and are skipped rather than stored. Returns `null` when no candidate
 * names a language we recognise, because an absent language is better than a
 * wrong one: it shows up as missing in the UI and can be corrected, whereas
 * `translated` looks like an answer.
 *
 * @param candidates - Tag names and/or column values, in any order
 */
export function resolveLanguageName(
  candidates: (string | null | undefined)[]
): CanonicalLanguage | null {
  const seen = new Set<string>()
  for (const raw of candidates) {
    if (!raw) continue
    // 'en-US' and 'zh_Hans' both reduce to their primary subtag.
    const norm = raw.toLowerCase().trim().split(/[-_]/)[0]
    if (norm) seen.add(norm)
  }
  if (seen.size === 0) return null

  // Priority order, not input order: 'translated, english' must give English,
  // and if a gallery ever carried both 'japanese' and 'english' the readable
  // language wins.
  for (const name of CANONICAL_LANGUAGES) {
    if (LANGUAGE_ALIASES[name].some((alias) => seen.has(alias))) return name
  }
  return null
}

File: kopibon-core/src/metadata/xml_utils.rs
//! Port of `src/main/services/xml-utils.ts` — escaping, entity decoding and
//! language mapping (07-metadata-spec §3 is the contract; the TS file is the
//! reference implementation and the cites below are its line numbers).

/// Characters that are illegal in XML 1.0 even when escaped
/// (xml-utils.ts:20). Stripped *before* entity substitution.
fn is_illegal_xml_char(c: char) -> bool {
    matches!(c, '\x00'..='\x08' | '\x0B' | '\x0C' | '\x0E'..='\x1F' | '\x7F')
}

/// Escape text for inclusion in XML element text or attribute values
/// (xml-utils.ts:28-36): illegal chars stripped first, then `& < > " '`,
/// ampersand first.
pub fn escape_xml(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for c in s.chars() {
        if is_illegal_xml_char(c) {
            continue;
        }
        match c {
            '&' => out.push_str("&amp;"),
            '<' => out.push_str("&lt;"),
            '>' => out.push_str("&gt;"),
            '"' => out.push_str("&quot;"),
            '\'' => out.push_str("&#39;"),
            _ => out.push(c),
        }
    }
    out
}

/// Decode XML character entities back to literal characters
/// (xml-utils.ts:47-56): numeric entities first (hex, then decimal order is
/// inherent to the scan), then the named five with `&` **last**.
///
/// JS `String.fromCodePoint` throws `RangeError: Invalid code point N` for
/// values above 0x10FFFF; that surfaces as the same error string here. JS can
/// build lone surrogates; Rust `String` cannot hold them (valid XML 1.0
/// excludes surrogate code points) so they decode to U+FFFD — no real input
/// in this library carries them.
pub fn decode_xml_entities(s: &str) -> Result<String, String> {
    let mut out = String::with_capacity(s.len());
    let mut chars = s.char_indices().peekable();
    while let Some((i, c)) = chars.next() {
        if c == '&' && s[i..].starts_with("&#") {
            match decode_numeric_entity(&s[i..], &mut out) {
                Ok(Some(len)) => {
                    // Advance past the entity: '&' is already consumed, and
                    // `len` is the whole entity's byte length.
                    for _ in 0..(len - 1) {
                        chars.next();
                    }
                    continue;
                }
                Ok(None) => {}
                Err(e) => return Err(e),
            }
        }
        out.push(c);
    }
    // Then the named five, ampersand last (xml-utils.ts:51-55).
    let out = out.replace("&lt;", "<");
    let out = out.replace("&gt;", ">");
    let out = out.replace("&quot;", "\"");
    let out = out.replace("&#39;", "'");
    let out = out.replace("&amp;", "&");
    Ok(out)
}

/// Decode one numeric entity at the start of `rest` (`&#x..;` or `&#..;`),
/// pushing its char onto `out`. Returns Ok(Some(byte length)) on success,
/// Ok(None) when `rest` does not start with a well-formed numeric entity
/// (the JS regexes simply fail to match then), Err for the JS RangeError.
fn decode_numeric_entity(rest: &str, out: &mut String) -> Result<Option<usize>, String> {
    let b = rest.as_bytes();
    // JS hex form is `&#x` — lowercase only (xml-utils.ts:49).
    let (hex, digits_start) = if b.len() > 3 && b[2] == b'x' {
        (true, 3)
    } else {
        (false, 2)
    };
    let mut end = digits_start;
    while end < b.len() {
        let d = b[end];
        let ok = if hex {
            d.is_ascii_hexdigit()
        } else {
            d.is_ascii_digit()
        };
        if !ok {
            break;
        }
        end += 1;
    }
    if end == digits_start || end >= b.len() || b[end] != b';' {
        return Ok(None);
    }
    let digits = &rest[digits_start..end];
    let value = if hex {
        match u128::from_str_radix(digits, 16) {
            Ok(v) => v,
            Err(_) => return Err(invalid_code_point_msg_js(digits, 16)),
        }
    } else {
        match digits.parse::<u128>() {
            Ok(v) => v,
            Err(_) => return Err(invalid_code_point_msg_js(digits, 10)),
        }
    };
    if value > 0x10FFFF {
        // JS: String.fromCodePoint throws RangeError: Invalid code point <n>
        return Err(format!(
            "Invalid code point {}",
            crate::metadata::js_number::js_to_string_u128(value)
        ));
    }
    match char::from_u32(value as u32) {
        Some(c) => out.push(c),
        None => out.push('\u{FFFD}'), // lone surrogate (see module doc)
    }
    Ok(Some(end + 1))
}

/// For absurdly long digit strings JS's `parseInt` yields Infinity and
/// `fromCodePoint` reports "Invalid code point Infinity"; unreachable for any
/// entity short enough to matter, but keep the message shape honest.
fn invalid_code_point_msg_js(digits: &str, radix: u32) -> String {
    let v = u128::from_str_radix(digits, radix).unwrap_or(u128::MAX);
    format!(
        "Invalid code point {}",
        crate::metadata::js_number::js_to_string_u128(v)
    )
}

// ─── Language mapping (xml-utils.ts:63-185) ─────────────────────────────────

/// `LANGUAGE_TO_ISO` (xml-utils.ts:79-108): nhentai tag names plus ISO 639-2
/// T and B forms, exactly as tabled.
const LANGUAGE_TO_ISO: &[(&str, &str)] = &[
    ("english", "en"),
    ("japanese", "ja"),
    ("chinese", "zh"),
    ("korean", "ko"),
    ("french", "fr"),
    ("spanish", "es"),
    ("german", "de"),
    ("italian", "it"),
    ("portuguese", "pt"),
    ("russian", "ru"),
    ("other", "ot"),
    ("eng", "en"),
    ("jpn", "ja"),
    ("zho", "zh"),
    ("chi", "zh"),
    ("kor", "ko"),
    ("fra", "fr"),
    ("fre", "fr"),
    ("spa", "es"),
    ("deu", "de"),
    ("ger", "de"),
    ("ita", "it"),
    ("por", "pt"),
    ("rus", "ru"),
];

/// Convert a free-text language value to an ISO 639-1 code
/// (xml-utils.ts:116-122). Bare two-letter lowercase passes through; the map
/// covers the rest; anything else (e.g. `translated`) is null.
pub fn to_iso_language(lang: Option<&str>) -> Option<String> {
    let lang = lang?;
    let lower = lang.to_lowercase();
    let lower = js_trim(&lower);
    if lower.is_empty() {
        return None;
    }
    if lower.len() == 2 && lower.bytes().all(|b| b.is_ascii_lowercase()) {
        return Some(lower.to_string());
    }
    LANGUAGE_TO_ISO
        .iter()
        .find(|(k, _)| *k == lower)
        .map(|(_, v)| v.to_string())
}

/// The only language values this app stores, in priority order
/// (xml-utils.ts:141).
pub const CANONICAL_LANGUAGES: [&str; 3] = ["English", "Japanese", "Chinese"];

const LANGUAGE_ALIASES: [(&str, &[&str]); 3] = [
    ("English", &["english", "en", "eng"]),
    ("Japanese", &["japanese", "ja", "jpn", "jp"]),
    ("Chinese", &["chinese", "zh", "zho", "chi", "cn"]),
];

/// Resolve language tags to one canonical name by **priority order, not input
/// order** (xml-utils.ts:166-185): each candidate reduces to its primary
/// subtag (`split(/[-_]/)[0]`), then the first canonical language whose
/// aliases are seen wins.
pub fn resolve_language_name(candidates: &[Option<String>]) -> Option<String> {
    let mut seen = std::collections::HashSet::new();
    for raw in candidates {
        let Some(raw) = raw else { continue };
        let lowered = raw.to_lowercase();
        let norm = js_trim(&lowered);
        // 'en-US' and 'zh_Hans' both reduce to their primary subtag.
        let norm = norm.split(['-', '_']).next().unwrap_or("");
        if !norm.is_empty() {
            seen.insert(norm.to_string());
        }
    }
    if seen.is_empty() {
        return None;
    }
    for (name, aliases) in LANGUAGE_ALIASES {
        if aliases.iter().any(|a| seen.contains(*a)) {
            return Some(name.to_string());
        }
    }
    None
}

/// JS `String.prototype.trim()` — trims the JS WhiteSpace + LineTerminator
/// set, which differs from Rust's `char::is_whitespace` in that it also drops
/// U+FEFF (and Rust's set is otherwise the same for practical input).
pub fn js_trim(s: &str) -> &str {
    s.trim_matches(|c: char| {
        matches!(
            c,
            '\u{0009}'..='\u{000D}'
                | '\u{0020}'
                | '\u{00A0}'
                | '\u{1680}'
                | '\u{2000}'..='\u{200A}'
                | '\u{2028}'
                | '\u{2029}'
                | '\u{202F}'
                | '\u{205F}'
                | '\u{3000}'
                | '\u{FEFF}'
        )
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn escape_order_and_illegal_strip() {
        assert_eq!(escape_xml("a<b&c\"d'e"), "a&lt;b&amp;c&quot;d&#39;e");
        assert_eq!(escape_xml("a\u{00}b\u{1F}c\u{7F}"), "abc");
        assert_eq!(escape_xml("\u{0B}"), "");
    }

    #[test]
    fn decode_named_then_numeric() {
        assert_eq!(decode_xml_entities("&lt;&amp;&gt;").unwrap(), "<&>");
        assert_eq!(decode_xml_entities("&#65;&#x41;").unwrap(), "AA");
        assert_eq!(decode_xml_entities("&amp;lt;").unwrap(), "&lt;");
        assert_eq!(decode_xml_entities("&lt;").unwrap(), "<");
        assert_eq!(
            decode_xml_entities("&notanentity;").unwrap(),
            "&notanentity;"
        );
        assert_eq!(decode_xml_entities("&#x1F600;").unwrap(), "\u{1F600}");
    }

    #[test]
    fn decode_range_error_string() {
        let err = decode_xml_entities("&#x110000;").unwrap_err();
        assert_eq!(err, "Invalid code point 1114112");
    }

    #[test]
    fn iso_and_names() {
        assert_eq!(to_iso_language(Some("Japanese")).as_deref(), Some("ja"));
        assert_eq!(to_iso_language(Some("jpn")).as_deref(), Some("ja"));
        assert_eq!(to_iso_language(Some("  ZH ")).as_deref(), Some("zh"));
        assert_eq!(to_iso_language(Some("translated")), None);
        assert_eq!(to_iso_language(Some("")), None);
        let cands = vec![Some("translated".into()), Some("english".into())];
        assert_eq!(resolve_language_name(&cands).as_deref(), Some("English"));
        let cands = vec![Some("japanese".into()), Some("english".into())];
        assert_eq!(resolve_language_name(&cands).as_deref(), Some("English"));
        let cands = vec![Some("japanese".into())];
        assert_eq!(resolve_language_name(&cands).as_deref(), Some("Japanese"));
        let cands = vec![Some("speechless".into())];
        assert_eq!(resolve_language_name(&cands), None);
    }
}
